UPGRADING-HADOOP.md

```diff
@@ -0,0 +1,181 @@
+
+1. finalize - cleans up the HDFS filesystem, and marks the rolling upgrade as
+complete
+2. downgrade - restores a previous version of hadoop (after which you ***must
+still finalize***)
+3. rollback - **not yet available** will restore a previous version of hadoop
+and restore HDFS file system to the state it was in when 'prepare' was run. 
+
+
+### Upgrade (standalone)
+
+  While we do not officially support the upgrade of a standalone deployment, it is possible using the following steps:
+
+  1. Config set version
+  2. Prepare rolling upgrade
+  3. Upgrade datanodes and then namenode using the standalone=true parameter
+
+
+### Finalize
+
+This is a fairly straight forward action and can be run on only one of the
+namenodes - this restarts the namenode daemons in 'normal' mode, but we
+recommend running it on both your namenodes to force those daemons to restart: 
+
+  juju action do namenode/$a hadoop-upgrade version=$version postupgrade=finalize
+
+One reason finalize should be run is that any files deleted after 'prepare' and
+before 'finalize' will remain (hidden) in your HDFS filesystem - therefore deleting
+files will not free up space - this is to enable a rollback. 
+
+
+### Downgrade
+
+Slightly more actions are required to downgrade HDFS than to finalize, and the
+recommended process is:
+
+1. Set hadoop_version config value for each service:
+
+
+    version=2.7.1
+    juju set namenode hadoop_version=$version
+    juju set slave hadoop_version=$version
+2. Downgrade datanodes:
+
+
+    juju action do slave/0 hadoop-upgrade version=$version postupgrade=downgrade
+    juju action do slave/1 hadoop-upgrade version=$version postupgrade=downgrade
+    juju action do slave/2 hadoop-upgrade version=$version postupgrade=downgrade
+    
+3. Downgrade both namenodes:
+
+
+    juju action do namenode/0 hadoop-upgrade version=$version postupgrade=downgrade
+    juju action do namenode/1 hadoop-upgrade version=$version postupgrade=downgrade
+
+***Once the downgrade is complete, you must still finalize the rolling upgrade***
+
+
+### Rollback
+
+Not yet supported - coming soon.
+
+
+### Action parameters
+
+    version:
+      type: string
+      description: destination hadoop version X.X.X, e.g. 2.7.2
+    prepare:
+      type: boolean
+      description: Must be run first - prepares fsimage backup
+      default: false
+    query:
+      type: boolean
+      description: Check if the fsimage backup is ready
+      default: false
+    postupgrade:
+      type: string
+      description: Should be specified after an upgrade
+      enum: [finalize, downgrade, rollback, rollback_finalize]
+    standalone:
+      type: boolean 
+      description: If False, upgrade an HA cluster without downtime. Otherwise perform standalone upgrade.
+      default: false
+    forceupgrade:
+      type: boolean
+      description: Doesn't check if the upgrade has been prepared - just download resource and switch symlink
+      default: false
+
+
+
+
+### To do
```

Comment:
Again, I do not like products that mention features they do not have and todos. It is like saying this production grade but... not yet production ready.... Lets put it in our backlog to support this in the future.